src/core/Oscillator.cpp

```diff
@@ -88,6 +118,206 @@ void Oscillator::update( sampleFrame * _ab, const fpp_t _frames,
+		for (float n = 1; n <= bands; n++)
+		{
+			m_generatedWaveTable[i] +=powf((float)-1.0, (float)(n + 1)) *
+					(1.0 / n) * sinf(F_2PI * i * n / (float)WAVETABLE_LENGTH);
+		}
+		max = fmax(max, m_generatedWaveTable[i]);
+	}
+
+	for (int i = 0; i < WAVETABLE_LENGTH; i++)
+	{
+		m_generatedWaveTable[i] /= max;
+	}
+}
+
+
+void Oscillator::generateTriangleWaveTable(int bands)
+{
+	float max = 0;
+	m_generatedWaveTable = new sample_t[WAVETABLE_LENGTH];
+	for (int i = 0 ; i < WAVETABLE_LENGTH; i++)
+	{
+		m_generatedWaveTable[i] = 0.0;
+		for (float n = 0; n <= bands * 0.5; n++)
+		{
+			m_generatedWaveTable[i] += powf((float)-1.0, (float)n) *
+					(1.0 / powf((float)(2 * n + 1),
+							   (float)2.0)) *
+					sinf(F_2PI * (2.0 * n + 1) * i / (float)WAVETABLE_LENGTH);
+		}
+		max = fmax(max, m_generatedWaveTable[i]);
+	}
+
+	for (int i = 0; i < WAVETABLE_LENGTH; i++)
+	{
+		m_generatedWaveTable[i] /= max;
+	}
+}
+
+void Oscillator::generateSquareWaveTable(int bands)
+{
+	float max = 0;
+	m_generatedWaveTable = new sample_t[WAVETABLE_LENGTH];
+	for (int i = 0; i < WAVETABLE_LENGTH; i++)
+	{
+		m_generatedWaveTable[i] = 0.0;
+		for (float n = 1; n <= bands; n += 2)
+		{
+			m_generatedWaveTable[i] += (1.0 / n) * sinf(F_2PI * i * n / WAVETABLE_LENGTH);
+		}
+		max = fmax(max, m_generatedWaveTable[i]);
+	}
+	for (int i = 0; i < WAVETABLE_LENGTH; i++)
+	{
+		m_generatedWaveTable[i] /= max;
+	}
+}
+
+
+
+
+//expects sample in sample buffer
+void Oscillator::generateFromFFT(int bands, float threshold)
+{
+	float max = 0;
+	
+	//set unrequired bands to zero 
+	for (int i = bands; i < WAVETABLE_LENGTH * 2 + 1 - bands; i++)
+	{
+		m_specBuf[i][0] = 0.0f;
+		m_specBuf[i][1] = 0.0f;
+	}
+	for (int i = 0; i < WAVETABLE_LENGTH * 2 + 1; ++i)
+	{
+		if (m_specBuf[i][0] * m_specBuf[i][0] + m_specBuf[i][1] * m_specBuf[i][1] < threshold)
+		{
+			m_specBuf[i][0] = 0.0f;
+			m_specBuf[i][1] = 0.0f;
+		}
+	}
+	//ifft
+	fftwf_execute(m_ifftPlan);
+	//normalise
+	for (int i = 0; i < WAVETABLE_LENGTH; ++i)
+	{
+		max = fmax(max, m_sampleBuffer[i]);
+	}
+	//copy to generateWaveTable
+	for (int i = 0; i < WAVETABLE_LENGTH; ++i)
+	{
+		m_generatedWaveTable[i] = m_sampleBuffer[i] / max;
+	}
+}
+
+
+int Oscillator::waveTableBandFromFreq(float freq)
+{
+	int i;
+	for (i = 0; i < s_waveTablesPerWaveformCount; ++i)
+	{
+		if (s_waveTableBandFreqs[i] > freq) return i;
+	}
+	return i;
+}
+
+sample_t ***Oscillator::s_waveTables = 0;
+float* Oscillator::s_waveTableBandFreqs = 0;
+int Oscillator::s_waveTablesPerWaveformCount = 0;
+
+
+
+void Oscillator::allocWaveTables()
+{
+	s_waveTables = new sample_t**[WaveShapes::NumWaveShapes];
+	memset(s_waveTables, 0, sizeof(sample_t) * WaveShapes::NumWaveShapes);
+	for (int i = 0; i < WaveShapes::NumWaveShapes; ++i)
+	{
+		s_waveTables[i] = new sample_t*[s_waveTablesPerWaveformCount +1];
+		memset(s_waveTables[i], 0, sizeof(sample_t) * s_waveTablesPerWaveformCount);
+	}
+}
+
+void Oscillator::createFFtPlans()
```

Comment:
Could you create something like `destroyFFTPlans` as well as this function?

Reply:
refactored as requested. I had spent too much time procrastinating this, had previously been a function,  I guess this solves the issue.